src/phongo_bson.c

```diff
@@ -790,37 +791,51 @@ static bool php_phongo_bson_visit_document(const bson_iter_t* iter ARG_UNUSED, c
 	php_phongo_bson_state* parent_state = (php_phongo_bson_state*) data;
 	php_phongo_bson_state  state;
 
-	if (!bson_iter_init(&child, v_document)) {
-		return false;
-	}
-
 	php_phongo_field_path_push(parent_state->field_path, key, PHONGO_FIELD_PATH_ITEM_DOCUMENT);
 
 	PHONGO_BSON_INIT_STATE(state);
 	php_phongo_bson_state_copy_ctor(&state, parent_state);
 
-	array_init(&state.zchild);
-
-	if (bson_iter_visit_all(&child, &php_bson_visitors, &state) || child.err_off) {
-		/* Iteration stopped prematurely due to corruption or a failed
-		 * visitor. Free state.zchild, which we just initialized, and return
-		 * true to stop iteration for our parent context. */
-		zval_ptr_dtor(&state.zchild);
-		php_phongo_bson_state_dtor(&state);
-		return true;
-	}
-
 	/* Check for entries in the fieldPath type map key, and use them to
 	 * override the default ones for this type */
 	php_phongo_handle_field_path_entry_for_compound_type(&state, &state.map.document);
 
+	/* Only traverse BSON document if we're not returning a raw BSON structure */
```

Comment:
Yes, `bson` takes priority over everything else. I can check to see if I can structure the visitors better, but I doubt it: since we not only allow `bson` to be specified as the root type, but also specifically for `array`, `document`, or any field path, we need to keep this logic in the array and document visitors as well.

Reply:
I realize we need logic in the document and array visitors, but I was originally asking if the "bson" handling could be front-loaded in the function to limit the diff. Something like:

```
if (state.map.array.type == PHONGO_TYPEMAP_BSON) {
  // copy the buffer and goto the end when we append to state.zchild
}
```

However, looking at the final function locally I don't think it's so bad so leave as-is.

Reply:
Feel free to resolve after reading.